### ISSUE TITLE
Exams: Export exam authorizations to Pearson

### ISSUE THREAD

**Issue by rhysyngsun** [mitodl/micromasters#2076]
In order to allow a user to schedule an exam we need to SFTP an exam authorization file (EAD) to Pearson containing rows for each user authorized. This is dependent on the preceding steps #1795 and #1797 having been completed for the given user's profile.

- [x] Data model to track exam authorization state
- [x] Method to create the EAD file locally
  - `ExamSeriesCode` will be tied to the program most likely *(waiting on verification)*
  - We **will** be using modules, they will be associated with the course. Waiting on confirmation of the values, but we we just store as an opaque string with maybe some light validation on the admin ui.
- [x] Celery task to generate and upload the EAD file

Open Questions:

- [x] What is the trigger on our end for a user being authorized? Do we currently track this somewhere?
- [x] I believe we need a flag on the `Course` record to signify that the course requires an exam. We also need to map the course to a Pearson exam code. Is there a case for not having these two be the same (i.e. presence of an exam code means the course has an exam)? I can't think of any. There's probably an argument for *not* putting this field on `Course` as well because it is Pearson-specific.
- [x] Pearson requires a range of eligibility dates (start/end) for authorizations. These can be updated, but we need a sensible default. This probably has some business logic implications (i.e. if someone took a course 5 years ago would we want to allow them to test for that?)
- [x] Are we supporting exam accommodations? If so, how is this expected to work? At the course level or per-user or both?

Deadline: 12/16/16

**Comment by Ferdi** [comment]
@rhysyngsun can you please add a bit of context / description ?

**Comment by rhysyngsun** [comment]
@Ferdi done, also added an open question that either you or @pdpinch may know the answer to.

**Comment by Ferdi** [comment]
@rhysyngsun thanks !  How is this file different from the file we generate here   https://github.com/mitodl/micromasters/issues/1795

**Comment by rhysyngsun** [comment]
@Ferdi #1795 is for the profile data only (CCD), this issue is for the exam authorizations (EAD) which give user X permission to schedule exam Y. Both files are required to be uploaded (and Pearson's response up to 12h later handled) in sequence in order to get to the state where a user could click a button and be able to navigate to Pearson's site and schedule an exam. There's a file in the directory Pearson shared with us named "MITx CCD-EAD via SFTP and SSO Flow.pdf" that diagrams the data flow at a high level.

**Comment by Ferdi** [comment]
Great , thanks ! It would be great if this info was part of the ticket description.  Saying "there is a  doc somewhere that only some people can view" defeat the purpose of having these discussions in the open.

**Comment by rhysyngsun** [comment]
@pdpinch questions listed in description above

@giocalitri would be interested in your thoughts on question 2 above, I'm leaning towards having a separate model, although a whole separate table for 1 column feels overkill.

**Comment by pdpinch** [comment]
Some (possible) answers:

1. User is authorized for an exam in a course when they have 1) passed the course and 2) paid for it. We will need a way to manually set the payment requirement

2. I agree that we need a flag. I think the presence of an exam code should be sufficient. However, I'm not sure if the exam code should be set on the course or the course run. I think it depends on how Pearson manages exam versions. 

3. We'll have to discuss the exam authorization dates with DEDP, but I think setting a range of today to a year from today should be a good default.

**Comment by rhysyngsun** [comment]
@pdpinch another question above

**Comment by pdpinch** [comment]
Is adding accommodations to the EAD file the only option for notifying Pearson of accommodation requirements? Let's ask Tommy when we meet today.

**Comment by rhysyngsun** [comment]
@pdpinch a few key points on accommodations from the docs/emails:

- Accommodation codes need to be setup with Pearson ahead of time
- Accommodation codes that haven't been setup will result in an error
- Once a candidate has registered, the accommodations on the authorization cannot be changed
- Thomas indicated that we could send accommodations separately from the EAD files (I assume this is some kind of manual process).

**Comment by rhysyngsun** [comment]
@pdpinch I'm not sure quite where we ended up with accommodation codes, I think you were going to talk to DEDP to see what they intended to do the first time around? Since I'm pretty much done with this issue aside from that, I'd suggest we deal with it in a separate issue.

**Comment by pdpinch** [comment]
I'm going to bring it up with DEDP in our meeting this afternoon.
